src/hooks.rs

```diff
@@ -0,0 +1,396 @@
+//! Lifecycle hooks.
+
+use std::fmt::Display;
+use std::process::ExitStatus;
+use std::str::FromStr;
+
+use clap::builder::ValueParserFactory;
+use clap::Arg;
+use clap::ArgAction;
+use clap::Args;
+use clap::FromArgMatches;
+use strum::EnumMessage;
+use strum::EnumProperty;
+use strum::IntoEnumIterator;
+use tokio::task::JoinHandle;
+
+use crate::ghci::GhciCommand;
+use crate::maybe_async_command::MaybeAsyncCommand;
+
+/// A lifecycle event that triggers hooks.
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    strum::EnumVariantNames,
+    strum::EnumIter,
+    strum::EnumMessage,
+    strum::EnumProperty,
+    strum::Display,
+    strum::AsRefStr,
+)]
+#[strum(serialize_all = "kebab-case")]
+pub enum LifecycleEvent {
+    /// When tests are run (after startup, after reloads).
+    #[strum(
+        message = "
+        Tests are run after startup and after reloads.
+        ",
+        props(help_name = "tests")
+    )]
+    Test,
+    /// When a `ghci` session is started (at `ghciwatch` startup and after restarts).
+    #[strum(message = "
+        Startup hooks run when `ghci` is started (at `ghciwatch` startup and after `ghci` restarts).
+    ")]
+    Startup,
+    /// When a module is changed or added.
+    #[strum(message = "
+        Reload hooks are run when modules are changed on disk.
+    ")]
+    Reload,
+    /// When a `ghci` session is restarted (when a module is removed or renamed).
+    #[strum(message = "
+        `ghci` is restarted when modules are removed or renamed.
+        See: https://gitlab.haskell.org/ghc/ghc/-/issues/11596
+    ")]
+    Restart,
+}
+
+impl LifecycleEvent {
+    fn supported_when(&self) -> Vec<When> {
+        match self {
+            LifecycleEvent::Test => vec![When::During],
+            LifecycleEvent::Startup | LifecycleEvent::Reload | LifecycleEvent::Restart => {
+                vec![When::Before, When::After]
+            }
+        }
+    }
```

Comment:
I think rather than check at runtime that we don't have an invalid combination of `LifecycleEvent` and `When` we should enforce that statically by combining them into a single data structure using something like this:

```rust
pub enum Bounded {
    Before,
    After,
}

pub enum LifecycleEvent {
    Test,
    StartUp{ when: When },
    Reload{ when: When },
    Restart{ when: When },
}
```

Reply:
I did consider this, but it gets kind of clumsy; for one, I'd have to add an explicit list of each possible event:

```rust
fn events() -> Vec<LifecycleEvent> {
  vec![
    LifecycleEvent::Test,
    LifecycleEvent::StartUp(When::Before),
    LifecycleEvent::StartUp(When::After),
    // ...
  ]
}
```

With that approach, when I add a new lifecycle event, nothing will tell me to add the new event to the list, and I also have to make sure to include every possible variant.

Reply:
Wouldn't that be an issue even for the existing approach?

Reply:
`supported_when` (the function you highlighted) matches on a value of type `LifecycleEvent`. The pattern match would become non-exhaustive.

Reply:
And I have a proc macro that generates a function which iterates over all the enum variants, which I can only use if they don't contain data.

Reply:
it is true that in that one case you might lose exhaustiveness, but on the balance I feel like you gain better safety/exhaustiveness by making invalid states unrepresentable

Reply:
Consider that the impetus of this system is that I [actually already forgot to fill in new entries in the old lifecycle hooks list](https://github.com/MercuryTechnologies/ghciwatch/pull/161)! This system prevents that. If there's a way to implement this data type that makes invalid states unrepresentable _and_ doesn't require me to explicitly list all the valid states, I'll gladly implement it.

Reply:
Can you clarify which list you're referring to?  I understand what you're asking for in the abstract but couldn't place where in each pull request this corresponds to